File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        m, n = len(heights), len(heights[0])
        p, a = set(), set()
        def dfs(i, j, v, ph):
            if i < 0 or j < 0 or i >= m or j >= n or (i,j) in v or heights[i][j] < ph:
                return 
            
            v.add((i,j))
            dfs(i-1, j, v, heights[i][j])
            dfs(i, j-1, v, heights[i][j])            
            dfs(i+1, j, v, heights[i][j])            
            dfs(i, j+1, v, heights[i][j])
            
        for c in range(n):
            dfs(0, c, p, heights[0][c])
            dfs(m-1, c, a, heights[m-1][c])  
            
        for r in range(m):
            dfs(r, 0, p, heights[r][0])
            dfs(r, n-1, a, heights[r][n-1])        
            
        res = []
        for i in range(m):
            for j in range(n):
                if (i,j) in p and (i,j) in a:
                    res.append((i,j))
                    
        return res
            

File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.java
class Solution {
    public record Pos(int x, int y) {
    }

    void dfs(int i, int j, Set<Pos> visSet, int [][] heights, int prevHeight) { 
        int n = heights.length, m = heights[0].length; 
        if (i < 0 || i >= n || j < 0 || j >= m) {
            return;
        } 
        Pos thisPos = new Pos(i, j);
        if (visSet.contains(thisPos) 
        ||  prevHeight > heights[i][j]) {
            return;
        }

        visSet.add(thisPos);
        dfs(i+1, j, visSet, heights, heights[i][j]);
        dfs(i, j+1, visSet, heights, heights[i][j]);
        dfs(i-1, j, visSet, heights, heights[i][j]);
        dfs(i, j-1, visSet, heights, heights[i][j]);
    }
    

    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        List<List<Integer>> res = new ArrayList<>();
        Set<Pos> pacific = new HashSet<>();
        Set<Pos> atlantic = new HashSet<>();
        int n = heights.length, m = heights.length; 
        for (int i = 0; i < n; i++) {
            dfs(i, 0, pacific, heights, Integer.MIN_VALUE);
            dfs(i, m-1, atlantic, heights, Integer.MIN_VALUE);
        }

        for (int j = 0; j < m; j++) {
            dfs(0, j, pacific, heights, Integer.MIN_VALUE);
            dfs(n-1, j, atlantic, heights, Integer.MIN_VALUE);
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                Pos thisPos = new Pos(i,j);
                if (pacific.contains(thisPos) && atlantic.contains(thisPos)) {
                    res.add(List.of(i, j));
                }

            }
        }

        return res;
    }
}